hoodie-client/src/main/java/com/uber/hoodie/func/BufferedIterator.java

```diff
@@ -0,0 +1,224 @@
+  private final Iterator<T> internalIterator;
+
+  private final AtomicReference<Exception> hasFailed = new AtomicReference(null);
+  private final AtomicBoolean isDone = new AtomicBoolean(false);
+  private T nextRecord;
+
+  private final AtomicLong readCounter = new AtomicLong(0);
+
+  private final AtomicLong samplingRecordCounter = new AtomicLong(-1);
+
+  public BufferedIterator(final Iterator<T> iterator, HoodieWriteConfig config) {
+    this.internalIterator = iterator;
+    this.bufferMemoryLimit = config.getInsertWriteBufferLimit();
+    this.schema = HoodieIOHandle.createHoodieWriteSchema(config);
+  }
+
+  private void adjustBufferSize(final T record) throws InterruptedException {
+    if (this.samplingRecordCounter.incrementAndGet() % RECORD_SAMPLING_RATE != 0) {
+      return;
+    }
+    final long recordSize = SizeEstimator.estimate(record);
+    final long newAvgRecordSize =
+      Math.max(1, (avgRecordSizeOfSampledRecords * numOfSampledRecords + recordSize) / (numOfSampledRecords + 1));
+    final int newRateLimit =
+      (int) Math.min(RECORD_CACHING_LIMIT, Math.max(1, this.bufferMemoryLimit / newAvgRecordSize));
+
+    if (newRateLimit > currentRateLimit) {
+      rateLimiter.release(newRateLimit - currentRateLimit);
+    } else if (newRateLimit < currentRateLimit) {
+      rateLimiter.acquire(currentRateLimit - newRateLimit);
+    }
+    currentRateLimit = newRateLimit;
+    avgRecordSizeOfSampledRecords = newAvgRecordSize;
+    numOfSampledRecords++;
+  }
+
+  private void insertRecord(T t) throws Exception {
+    adjustBufferSize(t);
+    rateLimiter.acquire();
+    // New Hoodie record is created to ensure that all the computation happens in reader thread.
+    HoodieRecordPayload payload = t.getData();
+    HoodieRecord newRecord = new HoodieRecord<>(t.getKey(),
+      new BufferedInsertPayload(payload.getInsertValue(schema), payload.getMetadata()));
+    buffer.put(new BufferedIteratorRecord(Optional.of(newRecord)));
+  }
+
+  private void readNextRecord() {
+    if (readCounter.incrementAndGet() % 1024 == 0) {
+      logger.debug("oldRateLimit:" + currentRateLimit + ":readCounter:" + readCounter.get()
+        + ":currentSize:" + buffer.size() + ":avgRecordSize:" + avgRecordSizeOfSampledRecords);
+    }
+    rateLimiter.release();
+    BufferedIteratorRecord<T> newRecord;
+    while (true) {
+      try {
+        throwExceptionIfFailed();
+        newRecord = buffer.poll(5, TimeUnit.SECONDS);
+        if (newRecord != null) {
+          break;
+        }
+      } catch (InterruptedException e) {
+        logger.error("error reading records from BufferedIterator", e);
+        throw new HoodieException(e);
+      }
+    }
+    if (newRecord.record.isPresent()) {
+      this.nextRecord = newRecord.record.get();
+    } else {
+      // We are done reading all the records from internal iterator.
+      this.isDone.set(true);
+      this.nextRecord = null;
+    }
+  }
+
+  public void startRecordReader() {
```

Comment:
please rename to something like `startBuffering()`, current name makes it sound like you are starting another thread

Reply:
makes sense. done.